plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java

```diff
@@ -16,67 +16,118 @@
+
     @Override
     protected void setup(Binder binder)
     {
         configBinder(binder).bindConfig(MappingConfig.class);
         binder.bind(DefaultIdentifierMapping.class).in(SINGLETON);
-        install(installModuleIf(
-                MappingConfig.class,
-                MappingConfig::isCaseInsensitiveNameMatching,
-                new CachingIdentifierMappingModule(),
-                new DefaultIdentifierMappingModule()));
-    }
 
-    private static final class CachingIdentifierMappingModule
-            extends AbstractConfigurationAwareModule
-    {
-        @Override
-        protected void setup(Binder binder)
-        {
+        MappingConfig config = buildConfigObject(MappingConfig.class);
+        if (config.isCaseInsensitiveNameMatching()) {
             Provider<JdbcClient> baseJdbcClientProvider = binder.getProvider(Key.get(JdbcClient.class, ForBaseJdbc.class));
             binder.bind(BaseJdbcClient.class).toProvider(() -> (BaseJdbcClient) baseJdbcClientProvider.get());
             binder.bind(IdentifierMapping.class)
                     .annotatedWith(ForCachingIdentifierMapping.class)
                     .to(DefaultIdentifierMapping.class)
                     .in(SINGLETON);
-            binder.bind(IdentifierMapping.class).to(CachingIdentifierMapping.class).in(SINGLETON);
+            binder.bind(IdentifierMapping.class)
+                    .annotatedWith(ForRuleBasedIdentifierMapping.class)
+                    .to(CachingIdentifierMapping.class)
+                    .in(SINGLETON);
+        }
+        else {
+            binder.bind(IdentifierMapping.class)
+                    .annotatedWith(ForRuleBasedIdentifierMapping.class)
+                    .to(DefaultIdentifierMapping.class)
+                    .in(SINGLETON);
+        }
+
+        if (config.getCaseInsensitiveNameMatchingConfigFile() != null) {
+            install(new RuleBasedIdentifierMappingModule());
+        }
+        else {
+            binder.bind(IdentifierMapping.class)
+                    .to(Key.get(IdentifierMapping.class, ForRuleBasedIdentifierMapping.class))
+                    .in(SINGLETON);
         }
     }
 
-    private static final class DefaultIdentifierMappingModule
+    private static final class RuleBasedIdentifierMappingModule
             extends AbstractConfigurationAwareModule
     {
         @Override
-        protected void setup(Binder binder)
+        protected void setup(Binder binder) {}
+
+        @Singleton
+        @Provides
+        public IdentifierMapping getIdentifierMapping(
+                CatalogName catalogName,
+                @ForRuleBasedIdentifierMapping IdentifierMapping identifierMapping,
+                MappingConfig config)
         {
-            binder.bind(IdentifierMapping.class).to(DefaultIdentifierMapping.class).in(SINGLETON);
+            String configFile = config.getCaseInsensitiveNameMatchingConfigFile();
```

Comment:
`checkArgument` instead? Having `configFile==null` is not possible according to `if` in `IdentifierMappingModule.setup`

Reply:
Also can we use Optional instead nullable return value of `config.getCaseInsensitiveNameMatchingConfigFile`?